src/test-run/index.js

```diff
@@ -0,0 +1,142 @@
+import delay from '../utils/delay';
+import MESSAGE from '../errors/runtime/message';
+import { GeneralError } from '../errors/runtime';
+
+
+export default class TestRun extends Session {
+    constructor (test, browserConnection, screenshotCapturer, opts) {
+        var uploadsRoot = path.dirname(test.fixture.path);
+
+        super(uploadsRoot);
+
+        this.opts               = opts;
+        this.test               = test;
+        this.browserConnection  = browserConnection;
+        this.screenshotCapturer = screenshotCapturer;
+
+        this.started        = false;
+        this.pendingCommand = null;
+        this.pendingRequest = null;
+        this.pendingJsError = null;
+
+        this.errs = [];
+    }
+
+    _getPayloadScript () {
+        // TODO
+    }
+
+    _getIframePayloadScript () {
+        // TODO
+    }
+
+    async _start () {
+        this.started = true;
+
+        try {
+            await this.test.fn(this);
+        }
+        catch (err) {
+            this._fatalError(err);
+        }
+        finally {
+            this._done();
+        }
+    }
+
+    _fatalError (err) {
+        this.errs.push(err);
+    }
+
+    _done () {
+        if (this.pendingRequest)
+            this._resolvePendingRequest(createDoneCommand());
+
+        if (this.pendingJsError)
+            this.errs.push(this.pendingJsError);
+
+        this.emit('done');
+    }
+
+    _resolvePendingCommand () {
+        this.pendingCommand.resolve();
+        this.pendingCommand = null;
+    }
+
+    _rejectPendingCommand (err) {
+        this.pendingCommand.reject(err);
+        this.pendingCommand = null;
+    }
+
+    _resolvePendingRequest (command) {
+        this.pendingRequest.resolve(command);
+        this.pendingRequest = null;
+    }
+
+    getAuthCredentials () {
+        // TODO
+    }
+
+    handleFileDownload () {
+        // TODO
+    }
+
+    handlePageError () {
+        // TODO
+    }
+
+    executeCommand (command) {
+        if (this.pendingCommand)
+            return Promise.reject(new GeneralError(MESSAGE.attemptToRunSeveralCommands));
+
+        var pendingJsError = this.pendingJsError;
+
+        if (pendingJsError) {
+            this.pendingJsError = null;
+            return Promise.reject(pendingJsError);
+        }
+
+        if (command.type === COMMAND_TYPE.wait)
+            return delay(command.arguments.ms);
+
+        if (this.pendingRequest)
+            this._resolvePendingRequest(command);
+
+        return new Promise((resolve, reject) => this.pendingCommand = { command, resolve, reject });
+    }
+}
+
+// Service message handlers
+var ServiceMessages = TestRun.prototype;
+
+ServiceMessages[CLIENT_MESSAGES.ready] = function (commandResult) {
+    this.pendingRequest = null;
+    
+    if (!this.started)
+        this._start();
+
+    if (this.pendingCommand) {
+        if (commandResult) {
+            if (commandResult.status === 'failed')
+                this._rejectPendingCommand(commandResult.err);
```

Comment:
As far as I understand currently it's misses branch that handles the situation, then we have pending command, but it was not accomplished and we need to tell client run it once again.